File: Strings/CheckPalindromeString.java
class Solution {
    public boolean isPalindrome(String s) {

        //Recursive approach
        // String s=str.replaceAll("[^A-Za-z0-9]", "").toUpperCase();

        // if(s.length()==0) return true;
        // if(s.length()==1) return true;
        // if(s.length()==2) {
        //     return s.charAt(0)==s.charAt(1)? true:false;
        // }
        // if (s.charAt(0) == s.charAt(s.length() - 1)) {
        // return isPalindrome(s.substring(1, s.length() - 1));
        // }
        // return false; 

        //2Poiner approach is optimized solution
        int i = 0;
        int j = s.length() - 1;
        while (i < j) {
            while (i < j && !Character.isLetterOrDigit(s.charAt(i))) {
                i++;
            }
            while (i < j && !Character.isLetterOrDigit(s.charAt(j))) {
                j--;
            }
            if (Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j))) {
                return false;
            }
            i++;
            j--;

        }
        return true;
    }
}

File: Strings/CheckPalindromeString.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    
    let i = 0;
    let j = s.length - 1;

    while (i < j) {
        while (i < j && !isLetterOrDigit(s[i])) {
            i++;
        }
        while (i < j && !isLetterOrDigit(s[j])) {
            j--;
        }
        if (i < j && s[i].toLowerCase() !== s[j].toLowerCase()) {
            return false;
        }
        i++;
        j--;
    }
    return true;
}

function isLetterOrDigit(c) {
    return /[a-zA-Z0-9]/.test(c);
    // The .test() method is called on the regular expression to test whether the string c contains any character that matches the pattern
    // defined by the regular expression.
}

